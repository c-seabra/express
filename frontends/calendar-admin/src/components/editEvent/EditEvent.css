/* Form */
.form {
  width: 100%;
  display: flex;
  flex-wrap: wrap;
  align-items: flex-start;
  justify-content: flex-start;
}

.formInput {
  margin-bottom: 1.5rem;
  display: flex;
  flex-wrap: wrap;
  align-items: flex-start;
  justify-content: space-between;
}

.formInput svg {
  width: 20px;
}

.formInput input {
  width: calc(100% - 42px);
  padding: .5rem .7rem;
}

.formEditContent {
  width: 450px;
  max-width: 100%;
  margin-right: 3rem;
}

.formEditInvitee {
  width: 250px;
  max-width: 100%;
}

.button {
  color: #3E3D44;
  border: 1px solid #cecece;
  transition: all .4s ease;
  padding: .5rem 1.2rem;
  background: white;
  border-radius: 4px;
  margin-left: 5px;
}

.button:hover {
  background: #FA7E23;
  color: #fff;
  cursor: pointer;
}

/* END Form */
