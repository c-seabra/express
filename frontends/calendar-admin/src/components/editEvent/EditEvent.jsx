import ButtonComponent from '@websummit/components/src/atoms/Button';
import { DestructiveButton } from '@websummit/components/src/atoms/Button';
import React, { useContext, useEffect, useState } from 'react';

import AddAttendance from '../addAttendance/AddAttendance';
import { DetailsContext } from '../calendar/Context';
import {
  Form,
  FormInput,
  FormLabel,
  FormSelect,
  FormTextArea,
  FormWrapper,
  InviteesListItem,
  Overlay,
  OverlayButtons,
  RemoveButton,
  StyledButton,
} from './EditEvent.styled';

const EditEvent = ({
  eventId,
  setEditPopupActive,
  title,
  location,
  description,
  startsAt,
  endsAt,
  eventFormatId,
  formats,
}) => {
  const {
    locations,
    rsvps,
    onDeleteEvent,
    onDeleteEventInvitation,
    onUpdateEvent,
  } = useContext(DetailsContext);
  const { onCreateEventInvitation } = useContext(DetailsContext);
  const [selections, setSelections] = useState([]);
  const [deletions, setDeletions] = useState([]);
  const [editedEvent, setEditedEvent] = useState({});
  const [deletePopupActive, setDeletePopupActive] = useState(false);
  const [locationName, setLocationName] = useState(location.name);
  const [formatName, setFormatName] = useState();
  const locationNames = locations.map((loc) => {
    return loc.name;
  });

  useEffect(() => {
    if (eventFormatId) {
      const { label } = formats.find((item) => item.id === eventFormatId);
      setFormatName(label);
    }
    if (location.id) {
      const { name } = locations.find((item) => item.id === location.id);
      setLocationName(name);
    }
  }, [location, locations, eventFormatId, formats]);

  const handleSetEditedEvent = (name, value) => {
    setEditedEvent((entity) => ({
      ...entity,
      [name]: value,
    }));
  };

  const handleLocationChange = (e) => {
    const { value } = e.target;
    if (locationNames.includes(value)) {
      setLocationName(value);
      const { id } = locations.find((item) => item.name === value);
      handleSetEditedEvent('location_id', id);
    } else {
      setLocationName(value);
      handleSetEditedEvent('location_name', value);
    }
  };

  const handleFormatChange = (e) => {
    const { value } = e.target;
    setFormatName(value);
    const { id } = formats.find((item) => item.label === value);
    handleSetEditedEvent('event_format_id', id);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (
      Object.keys(editedEvent).length !== 0 &&
      editedEvent.constructor === Object
    )
      onUpdateEvent(eventId, editedEvent);
    selections?.forEach((att) => {
      onCreateEventInvitation(eventId, att.id);
    });
    deletions?.forEach((att) => {
      onDeleteEventInvitation(eventId, att);
    });
    setEditPopupActive(false);
  };

  return (
    <>
      <DestructiveButton
        style={{
          marginTop: '1em',
          marginLeft: '75%',
        }}
        onClick={(e) => {
          e.preventDefault();
          setDeletePopupActive(true);
        }}
      >
        Delete Event
      </DestructiveButton>
      <Form onSubmit={(e) => handleSubmit(e)}>
        <FormWrapper>
          <FormLabel>Title: </FormLabel>
          <FormInput
            type="text"
            value={editedEvent.title !== undefined ? editedEvent.title : title}
            onChange={(e) => handleSetEditedEvent('title', e.target.value)}
          />
        </FormWrapper>
        <FormWrapper>
          <FormLabel>Starts At: </FormLabel>
          <FormInput
            type="datetime-local"
            value={
              editedEvent.starts_at !== undefined
                ? editedEvent.starts_at
                : startsAt
            }
            onChange={(e) => handleSetEditedEvent('starts_at', e.target.value)}
          />
        </FormWrapper>
        <FormWrapper>
          <FormLabel>Ends At: </FormLabel>
          <FormInput
            type="datetime-local"
            value={
              editedEvent.ends_at !== undefined ? editedEvent.ends_at : endsAt
            }
            onChange={(e) => handleSetEditedEvent('ends_at', e.target.value)}
          />
        </FormWrapper>
        <FormWrapper>
          <FormLabel htmlFor="location">Location: </FormLabel>
          <FormInput
            id="location"
            list="locations"
            type="text"
            value={locationName}
            onChange={(e) => handleLocationChange(e)}
          />
          <datalist id="locations">
            {locations?.map((loc, i) => (
              <option key={i} value={loc.name}>
                {loc.name}
              </option>
            ))}
          </datalist>
        </FormWrapper>
        <FormWrapper>
          <FormLabel htmlFor="event_format_id">Format: </FormLabel>
          <FormSelect
            id="event_format_id"
            type="text"
            value={formatName}
            onChange={(e) => handleFormatChange(e)}
          >
            <option defaultChecked>Please select a format</option>
            {formats?.map((format, i) => (
              <option key={i} value={format.label}>
                {format.label}
              </option>
            ))}
          </FormSelect>
        </FormWrapper>
        <FormWrapper>
          <FormLabel>Description: </FormLabel>
          <FormTextArea
            value={
              editedEvent.description !== undefined
                ? editedEvent.description
                : description
            }
            onChange={(e) =>
              handleSetEditedEvent('description', e.target.value)
            }
          />
        </FormWrapper>
        <FormWrapper>
          <FormLabel>Add attendances: </FormLabel>
          <AddAttendance
            ids={rsvps.map((r) => r.attendance.data.id)}
            selections={selections}
            setSelections={setSelections}
          />
        </FormWrapper>
        <div>
          {rsvps.map(
            (rsvp) =>
              !deletions.includes(rsvp.invitation.id) && (
                <InviteesListItem key={rsvp.invitation.id}>
                  <span style={{ width: '90%' }}>
                    {`${String(
                      rsvp.attendance.data.person.first_name,
                    )} ${String(rsvp.attendance.data.person.last_name)}`}
                  </span>
                  <RemoveButton
                    onClick={() => {
                      setDeletions([...deletions, rsvp.invitation.id]);
                    }}
                  />
                </InviteesListItem>
              ),
          )}
        </div>
        <FormWrapper>
          <StyledButton
            style={{
              backgroundColor: '#fff',
              color: '#0067e9',
              marginLeft: '65%',
            }}
            type="button"
            onClick={() => setEditPopupActive(false)}
          >
            Cancel
          </StyledButton>
          <StyledButton type="submit">Save</StyledButton>
        </FormWrapper>
        {deletePopupActive && (
          <Overlay>
            <h3>Are you sure you want to delete this event?</h3>
            <OverlayButtons>
              <ButtonComponent
                type="primary"
                onClick={() => {
                  onDeleteEvent(eventId);
                  setDeletePopupActive(false);
                }}
              >
                Yes
              </ButtonComponent>
              <ButtonComponent
                type="error"
                onClick={() => setDeletePopupActive(false)}
              >
                No
              </ButtonComponent>
            </OverlayButtons>
          </Overlay>
        )}
      </Form>
    </>
  );
};

export default EditEvent;
