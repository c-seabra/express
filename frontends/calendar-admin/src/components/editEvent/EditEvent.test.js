import React from 'react';
import renderer from 'react-test-renderer';

import EditEvent from './EditEvent';

describe('EditEvent', () => {
  let realUseContext;
  let useContextMock;

  beforeEach(() => {
    realUseContext = React.useContext;
    React.useContext = jest.fn();
    useContextMock = React.useContext;
  });

  afterEach(() => {
    React.useContext = realUseContext;
  });

  test('renders without props or context', () => {
    useContextMock.mockReturnValue({});
    const component = renderer.create(<EditEvent />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders with props, and no context', () => {
    useContextMock.mockReturnValue({});
    const component = renderer.create(
      <EditEvent
        description="Some event description"
        eventId="1234567"
        location={{
          name: 'Under palm tree',
          type: 'text',
        }}
        organizerId="1234567"
        setEditActive={false}
        title="Event Title"
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders with props and context', () => {
    useContextMock.mockReturnValue({
      locations: [
        {
          geography: { geometry: null, type: 'Feature' },
          id: 'czxczc',
          name: 'Sunset Summit',
        },
        {
          geography: { geometry: null, type: 'Feature' },
          id: 'dsadasd',
          name: 'Surf Summit (Invite-only)',
        },
        {
          geography: { geometry: null, type: 'Feature' },
          id: '123456778',
          name: 'FullSTK',
        },
      ],
      onDeleteEventInvitation () {
        return null;
      },
      onUpdateEvent () {
        return null;
      },
      rsvps: [
        {
          attendance: { data: { person: { first_name: 'Tomasz' } } },
          invitation: { id: 123, response: { response_status: 'accepted' } },
        },
      ],
    });
    const component = renderer.create(
      <EditEvent
        description="SOme event description"
        eventId="1234567"
        location={{
          name: 'Under palm tree',
          type: 'text',
        }}
        organizerId="1234567"
        setEditActive={false}
        title="Event Title"
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
