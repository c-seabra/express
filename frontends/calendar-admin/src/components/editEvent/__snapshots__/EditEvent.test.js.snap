// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditEvent renders with props and context 1`] = `
<form
  className={undefined}
  onSubmit={[Function]}
>
  <div
    className={undefined}
  >
    <div
      className={undefined}
    >
      <svg
        version="1.1"
        viewBox="0 0 15 15"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <title>
          Rectangle
        </title>
        <desc>
          Created with Sketch.
        </desc>
        <g
          fill="none"
          fillRule="evenodd"
          id="Symbols"
          stroke="none"
          strokeWidth="1"
        >
          <g
            fill="#FA7E23"
            fillRule="nonzero"
            id="1:1-Private-Meeting-small"
            transform="translate(-23.000000, -64.000000)"
          >
            <rect
              fill={undefined}
              height="15"
              id="Rectangle"
              rx="4"
              width="15"
              x="23"
              y="64"
            />
          </g>
        </g>
      </svg>
      <input
        name="title"
        onChange={[Function]}
        type="text"
        value="Event Title"
      />
    </div>
    <div
      className={undefined}
    >
      <input
        name="location_name"
        onChange={[Function]}
        type="text"
        value="Under palm tree"
      />
      <select
        name="location_id"
        onChange={[Function]}
        value={undefined}
      >
        <option
          value="czxczc"
        >
          Sunset Summit
        </option>
        <option
          value="dsadasd"
        >
          Surf Summit (Invite-only)
        </option>
        <option
          value="123456778"
        >
          FullSTK
        </option>
      </select>
    </div>
    <div
      className={undefined}
    >
      <textarea
        name="description"
        onChange={[Function]}
        value="SOme event description"
      />
    </div>
    <button
      className={undefined}
      type="submit"
    >
      Save
    </button>
    <button
      className={undefined}
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </div>
  <div
    className="undefined undefined"
  >
    <ul
      className={undefined}
    >
      <li>
        <div
          className={undefined}
        >
          <div
            className={undefined}
          >
            <object
              data={undefined}
              type="image/png"
            >
              <img
                alt="Invitee avatar"
                src="https://s3-eu-west-1.amazonaws.com/assets.cilabs.com/images/person.svg"
                style={
                  Object {
                    "borderColor": "transparent",
                  }
                }
              />
            </object>
            <svg
              viewBox="0 0 10 9"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs>
                <clipPath
                  id="vrmva"
                >
                  <path
                    d="M4.125 5.725L2.9 4.5l-.408.408 1.633 1.634 3.5-3.5-.408-.409z"
                  />
                </clipPath>
              </defs>
              <g>
                <g>
                  <g>
                    <path
                      d="M5 9a4.5 4.5 0 1 0 0-9 4.5 4.5 0 0 0 0 9z"
                      fill="#f8992c"
                    />
                  </g>
                  <g>
                    <g />
                    <g
                      clipPath="url(#vrmva)"
                    >
                      <g>
                        <path
                          d="M1.5 1h7v7h-7z"
                          fill="#fff"
                        />
                      </g>
                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </div>
          <div
            className={undefined}
          >
            <p
              className={undefined}
            >
              Tomasz undefined
            </p>
          </div>
          <span
            className={undefined}
            onClick={[Function]}
          >
            <svg
              viewBox="0 0 10 9"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs>
                <clipPath
                  id="en5za"
                >
                  <path
                    d="M7.333 2.637l-.47-.47L5 4.03 3.137 2.167l-.47.47L4.53 4.5 2.667 6.363l.47.47L5 4.97l1.863 1.863.47-.47L5.47 4.5z"
                  />
                </clipPath>
              </defs>
              <g>
                <g>
                  <g>
                    <path
                      d="M4.95 9a4.5 4.5 0 1 0 0-9 4.5 4.5 0 0 0 0 9z"
                      fill="#ed1846"
                    />
                  </g>
                </g>
                <g>
                  <g />
                  <g
                    clipPath="url(#en5za)"
                  >
                    <g>
                      <path
                        d="M1 .5h8v8H1z"
                        fill="#fff"
                      />
                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </span>
        </div>
      </li>
    </ul>
  </div>
</form>
`;

exports[`EditEvent renders with props, and no context 1`] = `
<form
  className={undefined}
  onSubmit={[Function]}
>
  <div
    className={undefined}
  >
    <div
      className={undefined}
    >
      <svg
        version="1.1"
        viewBox="0 0 15 15"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <title>
          Rectangle
        </title>
        <desc>
          Created with Sketch.
        </desc>
        <g
          fill="none"
          fillRule="evenodd"
          id="Symbols"
          stroke="none"
          strokeWidth="1"
        >
          <g
            fill="#FA7E23"
            fillRule="nonzero"
            id="1:1-Private-Meeting-small"
            transform="translate(-23.000000, -64.000000)"
          >
            <rect
              fill={undefined}
              height="15"
              id="Rectangle"
              rx="4"
              width="15"
              x="23"
              y="64"
            />
          </g>
        </g>
      </svg>
      <input
        name="title"
        onChange={[Function]}
        type="text"
        value="Event Title"
      />
    </div>
    <div
      className={undefined}
    >
      <input
        name="location_name"
        onChange={[Function]}
        type="text"
        value="Under palm tree"
      />
    </div>
    <div
      className={undefined}
    >
      <textarea
        name="description"
        onChange={[Function]}
        value="Some event description"
      />
    </div>
    <button
      className={undefined}
      type="submit"
    >
      Save
    </button>
    <button
      className={undefined}
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </div>
  <div
    className="undefined undefined"
  />
</form>
`;

exports[`EditEvent renders without props or context 1`] = `
<form
  className={undefined}
  onSubmit={[Function]}
>
  <div
    className={undefined}
  >
    <div
      className={undefined}
    >
      <svg
        version="1.1"
        viewBox="0 0 15 15"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <title>
          Rectangle
        </title>
        <desc>
          Created with Sketch.
        </desc>
        <g
          fill="none"
          fillRule="evenodd"
          id="Symbols"
          stroke="none"
          strokeWidth="1"
        >
          <g
            fill="#FA7E23"
            fillRule="nonzero"
            id="1:1-Private-Meeting-small"
            transform="translate(-23.000000, -64.000000)"
          >
            <rect
              fill={undefined}
              height="15"
              id="Rectangle"
              rx="4"
              width="15"
              x="23"
              y="64"
            />
          </g>
        </g>
      </svg>
      <input
        name="title"
        onChange={[Function]}
        type="text"
        value={undefined}
      />
    </div>
    <div
      className={undefined}
    >
      <input
        name="location_name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className={undefined}
    >
      <textarea
        name="description"
        onChange={[Function]}
        value={undefined}
      />
    </div>
    <button
      className={undefined}
      type="submit"
    >
      Save
    </button>
    <button
      className={undefined}
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </div>
  <div
    className="undefined undefined"
  />
</form>
`;
